core/pkg/ingress/types.go

```diff
@@ -0,0 +1,217 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"os/exec"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+
+	"k8s.io/ingress/core/pkg/ingress/annotations/auth"
+	"k8s.io/ingress/core/pkg/ingress/annotations/authreq"
+	"k8s.io/ingress/core/pkg/ingress/annotations/authtls"
+	"k8s.io/ingress/core/pkg/ingress/annotations/ipwhitelist"
+	"k8s.io/ingress/core/pkg/ingress/annotations/proxy"
+	"k8s.io/ingress/core/pkg/ingress/annotations/ratelimit"
+	"k8s.io/ingress/core/pkg/ingress/annotations/rewrite"
+	"k8s.io/ingress/core/pkg/ingress/defaults"
+)
+
+var (
+	// DefaultSSLDirectory defines the location where the SSL certificates will be generated
+	DefaultSSLDirectory = "/ingress-controller/ssl"
+)
+
+// Controller holds the methods to handle an Ingress backend
+type Controller interface {
+	// Start executes a command to start the backend.
+	// The command must run in foreground.
+	Start()
+	// Stop stops the backend
+	Stop() error
+	// Restart signal the backend with a new configuration file returning
+	// the combined output of Stdout and Stderr
+	Restart(data []byte) ([]byte, error)
+	// Tests returns the commands to execute that verifies if the configuration file is valid
+	// Example: nginx -t -c <file>
+	Test(file string) *exec.Cmd
+	// OnUpdate callback invoked from the sync queue https://k8s.io/ingress/core/blob/master/pkg/ingress/controller/controller.go#L387
+	// when an update occurs. This is executed frequently because Ingress
+	// controllers watches changes in:
+	// - Ingresses: main work
+	// - Secrets: referenced from Ingress rules with TLS configured
+	// - ConfigMaps: where the controller reads custom configuration
+	// - Services: referenced from Ingress rules and required to obtain
+	//	 information about ports and annotations
+	// - Endpoints: referenced from Services and what the backend uses
+	//	 to route traffic
+	//
+	// ConfigMap content of --configmap
+	// Configuration returns the translation from Ingress rules containing
+	// information about all the upstreams (service endpoints ) "virtual"
+	// servers (FQDN) and all the locations inside each server. Each
+	// location contains information about all the annotations were configured
+	// https://k8s.io/ingress/core/blob/master/pkg/ingress/types.go#L83
+	OnUpdate(*api.ConfigMap, Configuration) ([]byte, error)
+	// UpstreamDefaults returns the minimum settings required to configure the
+	// communication to upstream servers (endpoints)
+	UpstreamDefaults() defaults.Backend
+	// IsReloadRequired checks if the backend must be reloaded or not.
+	// The parameter contains the new rendered template
+	IsReloadRequired([]byte) bool
```

Comment:
> If feels weird that we have one method for checking if a reload is required, and another to test the config

Right. This is here just to avoid unnecessary reloads that can produce traffic service interruptions.  For nginx I use something like [`diff -u`](https://github.com/kubernetes/ingress/pull/15/files#diff-4b4d624221c37be9881c582a1c80b909R131) to compare the configurations
Maybe this can be the default?